### PR TITLE
[YUNIKORN-2525] dispatcher.Stop() waits an extra second unnecessarily

### PR DESCRIPTION
**Pull request by pbacsko**
### What is this PR for?
Speed up `dispatcher.Stop()` which also makes unit tests run faster.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2525

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/dispatcher/dispatcher.go [hunk 1]
```diff
@@ -257,13 +260,23 @@ func Stop() {
...
-		log.Log(log.ShimDispatcher).Info("waiting for dispatcher to be stopped",
-			zap.Int("remainingSeconds", maxTimeout))
-		time.Sleep(1 * time.Second)
-		maxTimeout--
+	stopWait := make(chan struct{})
+	defer close(stopWait)
+
+	go func() {
+		getDispatcher().stopped.Wait()
+		stopWait <- struct{}{}
```
Comment (chia7712) [inline pkg/dispatcher/dispatcher.go:268]: We can close the channel directly here.

**Review by chia7712 — APPROVED** [review]
overall +1 and one minor comment left.

File: pkg/dispatcher/dispatcher.go (hunk 1 above)
Comment (chia7712) [reply pkg/dispatcher/dispatcher.go:268]:
```go
	go func() {
		defer close(stopWait)
		getDispatcher().stopped.Wait()
	}()
```

Comment (pbacsko) [reply pkg/dispatcher/dispatcher.go:268]: That would mean double closure, which results in a panic. Closing inside `Stop()` seems to be a better place to me.

Comment (chia7712) [reply pkg/dispatcher/dispatcher.go:268]: > That would mean double closure

pardon me. why it causes double closure if we remove the `close(stopWait)` from `Stop()`.

```go
	stopWait := make(chan struct{})
        // defer close(stopWait) move this into goroutine

	go func() {
		defer close(stopWait)
		getDispatcher().stopped.Wait()
	}()

	// wait until the main event loop stops properly
	select {
	case <-stopWait:
		break
	case <-time.After(5 * time.Second):
		log.Log(log.ShimDispatcher).Info("dispatcher did not stop in time")
		break
	}
```

Comment (pbacsko) [reply pkg/dispatcher/dispatcher.go:268]: > pardon me. why it causes double closure if we remove the close(stopWait) from Stop().

In that case it doesn't. I just checked and a general principle is that channel shouldn't be closed on a receiver side. I don't think it's a big deal, but ok, I'll change it.

**Review by chia7712 — APPROVED** [review]

**Review by chenyulin0719 — APPROVED** [review]
+1 LGTM

**Pull request closed without merging**